File: Go/diff_utils/diff_utils_test.go
package diffutils

import (
	"strings"
	"testing"
)

// ========== DiffLines Tests ==========

func TestDiffLinesIdentical(t *testing.T) {
	a := "hello\nworld"
	result := DiffLines(a, a)

	if len(result.Diffs) != 2 {
		t.Errorf("Expected 2 diffs for identical strings, got %d", len(result.Diffs))
	}

	if result.Stats.Added != 0 || result.Stats.Removed != 0 {
		t.Errorf("Expected no changes for identical strings, got +added/-removed: %d/%d",
			result.Stats.Added, result.Stats.Removed)
	}

	for i, d := range result.Diffs {
		if d.Type != Equal {
			t.Errorf("Diff %d: expected Equal, got %v", i, d.Type)
		}
	}
}

func TestDiffLinesEmpty(t *testing.T) {
	result := DiffLines("", "")

	if len(result.Diffs) != 0 {
		t.Errorf("Expected 0 diffs for empty strings, got %d", len(result.Diffs))
	}
}

func TestDiffLinesInsertion(t *testing.T) {
	a := "line1\nline3"
	b := "line1\nline2\nline3"

	result := DiffLines(a, b)

	if result.Stats.Added != 1 {
		t.Errorf("Expected 1 addition, got %d", result.Stats.Added)
	}

	// Verify the inserted line is "line2"
	foundInsert := false
	for _, d := range result.Diffs {
		if d.Type == Insert && d.Text == "line2" {
			foundInsert = true
			break
		}
	}
	if !foundInsert {
		t.Error("Expected to find 'line2' as insertion")
	}
}

func TestDiffLinesDeletion(t *testing.T) {
	a := "line1\nline2\nline3"
	b := "line1\nline3"

	result := DiffLines(a, b)

	if result.Stats.Removed != 1 {
		t.Errorf("Expected 1 removal, got %d", result.Stats.Removed)
	}
}

func TestDiffLinesModification(t *testing.T) {
	a := "hello\nworld"
	b := "hello\ngo"

	result := DiffLines(a, b)

	if result.Stats.Removed != 1 || result.Stats.Added != 1 {
		t.Errorf("Expected 1 removal and 1 addition, got %d removed, %d added",
			result.Stats.Removed, result.Stats.Added)
	}
}

func TestDiffLinesUnicode(t *testing.T) {
	a := "你好\n世界"
	b := "你好\n世界！"

	result := DiffLines(a, b)

	if result.Stats.Removed != 1 || result.Stats.Added != 1 {
		t.Errorf("Expected modification for Unicode, got %d removed, %d added",
			result.Stats.Removed, result.Stats.Added)
	}
}

func TestDiffLinesMultipleChanges(t *testing.T) {
	a := "line1\nline2\nline3\nline4"
	b := "line0\nline1\nline3\nline5"

	result := DiffLines(a, b)

	if result.Stats.Added != 2 {
		t.Errorf("Expected 2 additions, got %d", result.Stats.Added)
	}
	if result.Stats.Removed != 2 {
		t.Errorf("Expected 2 removals, got %d", result.Stats.Removed)
	}
}

// ========== DiffChars Tests ==========

func TestDiffCharsIdentical(t *testing.T) {
	diffs := DiffChars("hello", "hello")

	for i, d := range diffs {
		if d.Type != Equal {
			t.Errorf("Diff %d: expected Equal, got %v", i, d.Type)
		}
	}
}

func TestDiffCharsInsert(t *testing.T) {
	diffs := DiffChars("ac", "abc")

	foundInsert := false
	for _, d := range diffs {
		if d.Type == Insert && d.Text == "b" {
			foundInsert = true
			break
		}
	}
	if !foundInsert {
		t.Error("Expected to find 'b' as insertion")
	}
}

func TestDiffCharsDelete(t *testing.T) {
	diffs := DiffChars("abc", "ac")

	foundDelete := false
	for _, d := range diffs {
		if d.Type == Delete && d.Text == "b" {
			foundDelete = true
			break
		}
	}
	if !foundDelete {
		t.Error("Expected to find 'b' as deletion")
	}
}

func TestDiffCharsReplace(t *testing.T) {
	diffs := DiffChars("cat", "bat")

	if len(diffs) < 2 {
		t.Error("Expected at least 2 diffs for replacement")
	}

	// First character should show deletion and insertion
	if diffs[0].Type != Delete || diffs[0].Text != "c" {
		t.Errorf("Expected first diff to be deletion of 'c', got %v '%s'", diffs[0].Type, diffs[0].Text)
	}
}

func TestDiffCharsUnicode(t *testing.T) {
	diffs := DiffChars("你好", "你坏")

	if len(diffs) == 0 {
		t.Error("Expected non-empty diff result")
	}

	// Should detect character change
	foundChange := false
	for _, d := range diffs {
		if d.Type != Equal {
			foundChange = true
			break
		}
	}
	if !foundChange {
		t.Error("Expected to detect Unicode character change")
	}
}

// ========== DiffWords Tests ==========

func TestDiffWordsIdentical(t *testing.T) {
	diffs := DiffWords("hello world", "hello world")

	for i, d := range diffs {
		if d.Type != Equal {
			t.Errorf("Diff %d: expected Equal, got %v", i, d.Type)
		}
	}
}

func TestDiffWordsInsert(t *testing.T) {
	diffs := DiffWords("hello world", "hello beautiful world")

	foundInsert := false
	for _, d := range diffs {
		if d.Type == Insert && d.Text == "beautiful" {
			foundInsert = true
			break
		}
	}
	if !foundInsert {
		t.Error("Expected to find 'beautiful' as insertion")
	}
}

func TestDiffWordsDelete(t *testing.T) {
	diffs := DiffWords("the quick brown fox", "the fox")

	if len(diffs) < 2 {
		t.Error("Expected multiple diffs for word deletion")
	}
}

// ========== LevenshteinDistance Tests ==========

func TestLevenshteinDistanceIdentical(t *testing.T) {
	d := LevenshteinDistance("hello", "hello")
	if d != 0 {
		t.Errorf("Expected distance 0 for identical strings, got %d", d)
	}
}

func TestLevenshteinDistanceEmpty(t *testing.T) {
	d := LevenshteinDistance("", "hello")
	if d != 5 {
		t.Errorf("Expected distance 5 for empty->hello, got %d", d)
	}

	d = LevenshteinDistance("hello", "")
	if d != 5 {
		t.Errorf("Expected distance 5 for hello->empty, got %d", d)
	}
}

func TestLevenshteinDistanceKittenSitting(t *testing.T) {
	// Classic example
	d := LevenshteinDistance("kitten", "sitting")
	if d != 3 {
		t.Errorf("Expected distance 3 for kitten->sitting, got %d", d)
	}
}

func TestLevenshteinDistanceSubstitution(t *testing.T) {
	d := LevenshteinDistance("cat", "bat")
	if d != 1 {
		t.Errorf("Expected distance 1 for cat->bat, got %d", d)
	}
}

func TestLevenshteinDistanceUnicode(t *testing.T) {
	d := LevenshteinDistance("你好", "你坏")
	if d != 1 {
		t.Errorf("Expected distance 1 for Unicode substitution, got %d", d)
	}
}

// ========== Similarity Tests ==========

func TestSimilarityIdentical(t *testing.T) {
	s := Similarity("hello", "hello")
	if s != 1.0 {
		t.Errorf("Expected similarity 1.0 for identical strings, got %f", s)
	}
}

func TestSimilarityEmpty(t *testing.T) {
	s := Similarity("", "hello")
	if s != 0.0 {
		t.Errorf("Expected similarity 0.0 for empty string, got %f", s)
	}
}

func TestSimilarityPartial(t *testing.T) {
	s := Similarity("hello", "hallo")
	// One character different out of 5
	expected := 0.8
	if s < expected-0.01 || s > expected+0.01 {
		t.Errorf("Expected similarity ~%f, got %f", expected, s)
	}
}

func TestSimilarityCompletelyDifferent(t *testing.T) {
	s := Similarity("abc", "xyz")
	// All different
	if s > 0.5 {
		t.Errorf("Expected low similarity for completely different strings, got %f", s)
	}
}

// ========== LCS Tests ==========

func TestLCSIdentical(t *testing.T) {
	lcs, length := LCS("ABC", "ABC")
	if lcs != "ABC" {
		t.Errorf("Expected LCS 'ABC', got '%s'", lcs)
	}
	if length != 3 {
		t.Errorf("Expected length 3, got %d", length)
	}
}

func TestLCSEmpty(t *testing.T) {
	lcs, length := LCS("", "ABC")
	if length != 0 {
		t.Errorf("Expected length 0 for empty string, got %d", length)
	}
	if lcs != "" {
		t.Errorf("Expected empty LCS, got '%s'", lcs)
	}
}

func TestLCSClassic(t *testing.T) {
	lcs, length := LCS("ABCBDAB", "BDCABA")
	// Possible LCS: "BCBA" or "BDAB" with length 4
	if length != 4 {
		t.Errorf("Expected LCS length 4, got %d (lcs: '%s')", length, lcs)
	}
}

func TestLCSUnicode(t *testing.T) {
	lcs, length := LCS("你好世界", "你好中国")
	if length < 2 {
		t.Errorf("Expected LCS length >= 2, got %d (lcs: '%s')", length, lcs)
	}
	if !strings.Contains(lcs, "你好") {
		t.Errorf("Expected LCS to contain '你好', got '%s'", lcs)
	}
}

// ========== FormatUnified Tests ==========

func TestFormatUnifiedNoChanges(t *testing.T) {
	result := DiffLines("line1\nline2", "line1\nline2")
	uni := FormatUnified(result, 3, "a.txt", "b.txt")

	if !strings.Contains(uni, "--- a.txt") || !strings.Contains(uni, "+++ b.txt") {
		t.Error("Expected unified diff headers")
	}
}

func TestFormatUnifiedWithChanges(t *testing.T) {
	result := DiffLines("line1\nline2", "line1\nmodified")
	uni := FormatUnified(result, 3, "a.txt", "b.txt")

	if !strings.Contains(uni, "--- a.txt") {
		t.Error("Expected --- header")
	}
	if !strings.Contains(uni, "+++ b.txt") {
		t.Error("Expected +++ header")
	}
	if !strings.Contains(uni, "@@") {
		t.Error("Expected hunk header @@")
	}
}

// ========== FormatColor Tests ==========

func TestFormatColor(t *testing.T) {
	diffs := []Diff{
		{Type: Equal, Text: "same"},
		{Type: Delete, Text: "old"},
		{Type: Insert, Text: "new"},
	}
	result := FormatColor(diffs)

	if !strings.Contains(result, "same") {
		t.Error("Expected 'same' in output")
	}
	if !strings.Contains(result, "old") {
		t.Error("Expected 'old' in output")
	}
	if !strings.Contains(result, "new") {
		t.Error("Expected 'new' in output")
	}
	// Should contain ANSI codes
	if !strings.Contains(result, "\033[") {
		t.Error("Expected ANSI color codes in output")
	}
}

// ========== FormatHTML Tests ==========

func TestFormatHTML(t *testing.T) {
	diffs := []Diff{
		{Type: Equal, Text: "same"},
		{Type: Delete, Text: "old"},
		{Type: Insert, Text: "new"},
	}
	result := FormatHTML(diffs)

	if !strings.Contains(result, "<div") {
		t.Error("Expected div wrapper")
	}
	if !strings.Contains(result, "<del>old</del>") {
		t.Error("Expected <del> tag for deletion")
	}
	if !strings.Contains(result, "<ins>new</ins>") {
		t.Error("Expected <ins> tag for insertion")
	}
	if !strings.Contains(result, "<span>same</span>") {
		t.Error("Expected <span> tag for equal text")
	}
}

func TestFormatHTMLEscape(t *testing.T) {
	diffs := []Diff{
		{Type: Equal, Text: "<script>alert('xss')</script>"},
	}
	result := FormatHTML(diffs)

	if strings.Contains(result, "<script>") {
		t.Error("HTML should be escaped")
	}
	if !strings.Contains(result, "&lt;script&gt;") {
		t.Error("Expected escaped HTML entities")
	}
}

// ========== Edge Cases ==========

func TestDiffLinesTrailingNewline(t *testing.T) {
	a := "line1\nline2"
	b := "line1\nline2\n"

	result := DiffLines(a, b)

	// Trailing newline should result in an insertion
	if result.Stats.Added != 1 {
		t.Errorf("Expected 1 addition for trailing newline, got %d", result.Stats.Added)
	}
}

func TestDiffCharsEmoji(t *testing.T) {
	// Test with multi-byte Unicode characters
	diffs := DiffChars("hello 👋 world", "hello 🌍 world")

	foundDelete := false
	foundInsert := false
	for _, d := range diffs {
		if d.Type == Delete && strings.Contains(d.Text, "👋") {
			foundDelete = true
		}
		if d.Type == Insert && strings.Contains(d.Text, "🌍") {
			foundInsert = true
		}
	}

	if !foundDelete || !foundInsert {
		t.Error("Expected emoji substitution to be detected")
	}
}

func TestLargeDiff(t *testing.T) {
	// Test with larger inputs
	var a, b strings.Builder
	for i := 0; i < 1000; i++ {
		a.WriteString("line\n")
		if i%2 == 0 {
			b.WriteString("line\n")
		} else {
			b.WriteString("modified\n")
		}
	}

	result := DiffLines(a.String(), b.String())

	if result.Stats.Removed != 500 || result.Stats.Added != 500 {
		t.Errorf("Expected 500 modifications, got %d removed, %d added",
			result.Stats.Removed, result.Stats.Added)
	}
}

// ========== Benchmark ==========

func BenchmarkDiffLines(b *testing.B) {
	a := strings.Repeat("line\n", 1000)
	c := strings.Repeat("line\n", 1000)
	for i := 0; i < b.N; i++ {
		DiffLines(a, c)
	}
}

func BenchmarkLevenshteinDistance(b *testing.B) {
	a := strings.Repeat("a", 1000)
	c := strings.Repeat("b", 1000)
	for i := 0; i < b.N; i++ {
		LevenshteinDistance(a, c)
	}
}

func BenchmarkDiffChars(b *testing.B) {
	a := strings.Repeat("hello", 100)
	c := strings.Repeat("world", 100)
	for i := 0; i < b.N; i++ {
		DiffChars(a, c)
	}
}

File: JavaScript/diff_utils/diff_utils_test.js
/**
 * diff_utils 测试文件
 */

const assert = require('assert');
const {
  DiffType,
  diffLines,
  diffChars,
  diffWords,
  mergeDiff,
  unifiedDiff,
  colorDiff,
  diffStats,
  applyDiff,
  reverseDiff,
  similarity,
  longestCommonSubsequence,
  longestCommonSubstring,
  diffToHtml,
  parseUnifiedDiff,
  threeWayMerge,
  computeLCSMatrix
} = require('./mod.js');

// 测试计数器
let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`✓ ${name}`);
    passed++;
  } catch (error) {
    console.log(`✗ ${name}`);
    console.log(`  Error: ${error.message}`);
    failed++;
  }
}

function testGroup(name) {
  console.log(`\n【${name}】`);
}

// ============== 测试开始 ==============

testGroup('基础差异计算');

test('diffLines - 完全相同的文本', () => {
  const text1 = 'hello\nworld';
  const text2 = 'hello\nworld';
  const diff = diffLines(text1, text2);
  
  assert.strictEqual(diff.length, 2);
  assert.strictEqual(diff[0].type, DiffType.EQUAL);
  assert.strictEqual(diff[0].value, 'hello');
  assert.strictEqual(diff[1].type, DiffType.EQUAL);
  assert.strictEqual(diff[1].value, 'world');
});

test('diffLines - 完全不同的文本', () => {
  const text1 = 'hello';
  const text2 = 'world';
  const diff = diffLines(text1, text2);
  
  assert.strictEqual(diff.length, 2);
  assert.strictEqual(diff[0].type, DiffType.DELETE);
  assert.strictEqual(diff[0].value, 'hello');
  assert.strictEqual(diff[1].type, DiffType.ADD);
  assert.strictEqual(diff[1].value, 'world');
});

test('diffLines - 部分不同', () => {
  const text1 = 'hello\nworld\nfoo';
  const text2 = 'hello\nbar\nfoo';
  const diff = diffLines(text1, text2);
  
  // 应该有 4 部分：equal(hello), delete(world), add(bar), equal(foo)
  assert.strictEqual(diff.length, 4);
  assert.strictEqual(diff[0].type, DiffType.EQUAL);
  assert.strictEqual(diff[0].value, 'hello');
  assert.strictEqual(diff[1].type, DiffType.DELETE);
  assert.strictEqual(diff[1].value, 'world');
  assert.strictEqual(diff[2].type, DiffType.ADD);
  assert.strictEqual(diff[2].value, 'bar');
  assert.strictEqual(diff[3].type, DiffType.EQUAL);
  assert.strictEqual(diff[3].value, 'foo');
});

test('diffLines - 添加行', () => {
  const text1 = 'hello\nworld';
  const text2 = 'hello\nbeautiful\nworld';
  const diff = diffLines(text1, text2);
  
  assert.ok(diff.some(d => d.type === DiffType.ADD && d.value === 'beautiful'));
});

test('diffLines - 删除行', () => {
  const text1 = 'hello\nbeautiful\nworld';
  const text2 = 'hello\nworld';
  const diff = diffLines(text1, text2);
  
  assert.ok(diff.some(d => d.type === DiffType.DELETE && d.value === 'beautiful'));
});

testGroup('字符级差异');

test('diffChars - 字符差异', () => {
  const text1 = 'hello';
  const text2 = 'hallo';
  const diff = diffChars(text1, text2);
  
  assert.ok(diff.some(d => d.type === DiffType.DELETE && d.value === 'e'));
  assert.ok(diff.some(d => d.type === DiffType.ADD && d.value === 'a'));
});

test('diffChars - 完全相同', () => {
  const diff = diffChars('abc', 'abc');
  // diffChars 按字符返回，需要用 mergeDiff 合并
  const merged = mergeDiff(diff);
  assert.strictEqual(merged.length, 1);
  assert.strictEqual(merged[0].type, DiffType.EQUAL);
  assert.strictEqual(merged[0].value, 'abc');
});

testGroup('单词级差异');

test('diffWords - 单词差异', () => {
  const text1 = 'hello world foo';
  const text2 = 'hello bar foo';
  const diff = diffWords(text1, text2);
  
  assert.ok(diff.some(d => d.type === DiffType.DELETE && d.value === 'world'));
  assert.ok(diff.some(d => d.type === DiffType.ADD && d.value === 'bar'));
});

testGroup('差异合并');

test('mergeDiff - 合并相邻相同类型', () => {
  const diff = [
    { type: DiffType.EQUAL, value: 'a' },
    { type: DiffType.EQUAL, value: 'b' },
    { type: DiffType.ADD, value: 'c' },
    { type: DiffType.ADD, value: 'd' }
  ];
  
  const merged = mergeDiff(diff);
  
  assert.strictEqual(merged.length, 2);
  assert.strictEqual(merged[0].value, 'ab');
  assert.strictEqual(merged[1].value, 'cd');
});

testGroup('统一格式 diff');

test('unifiedDiff - 生成统一格式', () => {
  const text1 = 'line1\nline2\nline3';
  const text2 = 'line1\nmodified\nline3';
  const result = unifiedDiff(text1, text2, {
    filename1: 'a/original.txt',
    filename2: 'b/new.txt',
    context: 1
  });
  
  assert.ok(result.includes('--- a/original.txt'));
  assert.ok(result.includes('+++ b/new.txt'));
  assert.ok(result.includes('@@'));
  assert.ok(result.includes('-line2'));
  assert.ok(result.includes('+modified'));
});

test('unifiedDiff - 多行变更', () => {
  const text1 = 'a\nb\nc\nd\ne';
  const text2 = 'a\nx\ny\nc\nd\ne';
  const result = unifiedDiff(text1, text2);
  
  assert.ok(result.includes('-b'));
  assert.ok(result.includes('+x'));
  assert.ok(result.includes('+y'));
});

testGroup('差异统计');

test('diffStats - 统计信息', () => {
  const diff = [
    { type: DiffType.EQUAL, value: 'a' },
    { type: DiffType.DELETE, value: 'bb' },
    { type: DiffType.ADD, value: 'ccc' }
  ];
  
  const stats = diffStats(diff);
  
  assert.strictEqual(stats.equal, 1);
  assert.strictEqual(stats.deleted, 1);
  assert.strictEqual(stats.added, 1);
  assert.strictEqual(stats.deletedChars, 2);
  assert.strictEqual(stats.addedChars, 3);
});

testGroup('差异应用与回滚');

test('applyDiff - 应用差异', () => {
  const text1 = 'hello\nworld';
  const text2 = 'hello\nbeautiful\nworld';
  const diff = diffLines(text1, text2);
  const result = applyDiff(text1, diff);
  
  assert.strictEqual(result, text2);
});

test('reverseDiff - 反转差异', () => {
  const diff = [
    { type: DiffType.ADD, value: 'a' },
    { type: DiffType.DELETE, value: 'b' },
    { type: DiffType.EQUAL, value: 'c' }
  ];
  
  const reversed = reverseDiff(diff);
  
  assert.strictEqual(reversed[0].type, DiffType.DELETE);
  assert.strictEqual(reversed[1].type, DiffType.ADD);
  assert.strictEqual(reversed[2].type, DiffType.EQUAL);
});

testGroup('相似度计算');

test('similarity - 完全相同', () => {
  const sim = similarity('hello', 'hello');
  assert.strictEqual(sim, 1);
});

test('similarity - 完全不同', () => {
  const sim = similarity('abc', 'xyz');
  assert.strictEqual(sim, 0);
});

test('similarity - 部分相同', () => {
  const sim = similarity('hello', 'hallo');
  assert.ok(sim > 0.5 && sim < 1);
});

test('similarity - 空字符串', () => {
  assert.strictEqual(similarity('', ''), 1);
  assert.strictEqual(similarity('', 'a'), 0);
  assert.strictEqual(similarity('a', ''), 0);
});

testGroup('最长公共子序列/子串');

test('longestCommonSubsequence - 基本测试', () => {
  const lcs = longestCommonSubsequence('ABCBDAB', 'BDCABA');
  assert.strictEqual(lcs.length, 4);
  // LCS 可能是 BCBA 或 BDAB
  assert.ok(['BCBA', 'BDAB', 'BCAB'].includes(lcs));
});

test('longestCommonSubsequence - 完全相同', () => {
  const lcs = longestCommonSubsequence('hello', 'hello');
  assert.strictEqual(lcs, 'hello');
});

test('longestCommonSubsequence - 无公共部分', () => {
  const lcs = longestCommonSubsequence('abc', 'xyz');
  assert.strictEqual(lcs, '');
});

test('longestCommonSubstring - 基本测试', () => {
  const lcs = longestCommonSubstring('ABCBDAB', 'BDCABA');
  // 最长公共子串应该是 AB 或 BD
  assert.ok(lcs.length >= 2);
});

test('longestCommonSubstring - 完全相同', () => {
  const lcs = longestCommonSubstring('hello', 'hello');
  assert.strictEqual(lcs, 'hello');
});

testGroup('HTML 输出');

test('diffToHtml - 生成 HTML', () => {
  const diff = [
    { type: DiffType.EQUAL, value: 'hello' },
    { type: DiffType.ADD, value: 'world' },
    { type: DiffType.DELETE, value: 'foo' }
  ];
  
  const html = diffToHtml(diff);
  
  assert.ok(html.includes('class="diff-equal"'));
  assert.ok(html.includes('class="diff-add"'));
  assert.ok(html.includes('class="diff-delete"'));
  assert.ok(html.includes('hello'));
  assert.ok(html.includes('world'));
});

test('diffToHtml - 转义特殊字符', () => {
  const diff = [
    { type: DiffType.EQUAL, value: '<script>alert("xss")</script>' }
  ];
  
  const html = diffToHtml(diff);
  
  assert.ok(!html.includes('<script>'));
  assert.ok(html.includes('&lt;'));
  assert.ok(html.includes('&gt;'));
});

testGroup('解析统一格式');

test('parseUnifiedDiff - 解析 diff', () => {
  const unified = `--- a/original
+++ b/new
@@ -1,3 +1,3 @@
 line1
-line2
+modified
 line3`;
  
  const diff = parseUnifiedDiff(unified);
  
  assert.ok(diff.some(d => d.type === DiffType.EQUAL && d.value === 'line1'));
  assert.ok(diff.some(d => d.type === DiffType.DELETE && d.value === 'line2'));
  assert.ok(diff.some(d => d.type === DiffType.ADD && d.value === 'modified'));
  assert.ok(diff.some(d => d.type === DiffType.EQUAL && d.value === 'line3'));
});

testGroup('三方合并');

test('threeWayMerge - 无冲突', () => {
  const base = 'line1\nline2\nline3';
  const ours = 'line1\nline2\nline3';
  const theirs = 'line1\nline2\nline3';
  
  const result = threeWayMerge(base, ours, theirs);
  
  assert.strictEqual(result.success, true);
  assert.strictEqual(result.conflicts.length, 0);
});

test('threeWayMerge - 检测冲突', () => {
  const base = 'line1\nline2\nline3';
  const ours = 'line1\nour-change\nline3';
  const theirs = 'line1\ntheir-change\nline3';
  
  const result = threeWayMerge(base, ours, theirs);
  
  // 检测到冲突
  assert.ok(result.diffOurs.some(d => d.type === DiffType.ADD || d.type === DiffType.DELETE));
  assert.ok(result.diffTheirs.some(d => d.type === DiffType.ADD || d.type === DiffType.DELETE));
});

testGroup('LCS 矩阵计算');

test('computeLCSMatrix - 基本测试', () => {
  const arr1 = ['A', 'B', 'C'];
  const arr2 = ['A', 'C'];
  const dp = computeLCSMatrix(arr1, arr2);
  
  // LCS 长度应该是 2 (A 和 C)
  assert.strictEqual(dp[3][2], 2);
});

test('computeLCSMatrix - 完全相同', () => {
  const arr = ['a', 'b', 'c'];
  const dp = computeLCSMatrix(arr, arr);
  
  assert.strictEqual(dp[3][3], 3);
});

test('computeLCSMatrix - 无公共元素', () => {
  const dp = computeLCSMatrix(['a', 'b'], ['c', 'd']);
  
  assert.strictEqual(dp[2][2], 0);
});

testGroup('颜色输出');

test('colorDiff - 包含 ANSI 颜色码', () => {
  const diff = [
    { type: DiffType.ADD, value: 'added' },
    { type: DiffType.DELETE, value: 'deleted' },
    { type: DiffType.EQUAL, value: 'equal' }
  ];
  
  const colored = colorDiff(diff);
  
  assert.ok(colored.includes('\x1b[32m')); // 绿色
  assert.ok(colored.includes('\x1b[31m')); // 红色
  assert.ok(colored.includes('\x1b[0m')); // 重置
});

testGroup('边界情况');

test('空字符串处理', () => {
  const diff = diffLines('', '');
  // 空字符串 split('\n') 返回 ['']，所以有 1 个空行元素
  // 这是正确的行为，空文件也有 1 个空行
  assert.ok(diff.length >= 1);
  assert.ok(diff.every(d => d.type === DiffType.EQUAL));
});

test('单行文本', () => {
  const diff = diffLines('hello', 'world');
  assert.strictEqual(diff.length, 2);
  assert.strictEqual(diff[0].type, DiffType.DELETE);
  assert.strictEqual(diff[1].type, DiffType.ADD);
});

test('大量文本', () => {
  const lines1 = Array(1000).fill('line').map((v, i) => v + i);
  const lines2 = [...lines1];
  lines2[500] = 'modified';
  
  const diff = diffLines(lines1.join('\n'), lines2.join('\n'));
  
  assert.ok(diff.length > 0);
  assert.ok(diff.some(d => d.value === 'modified'));
});

// ============== 测试结束 ==============

console.log('\n' + '='.repeat(50));
console.log(`测试完成: ${passed} 通过, ${failed} 失败`);
console.log('='.repeat(50));

if (failed > 0) {
  process.exit(1);
}